proposals/nnn-circular-buffer.md

```diff
@@ -0,0 +1,151 @@
+# Circular Buffer
+
+* Proposal: [SE-CircularBuffer](NNNN-CircularBuffer.md)
+* Authors: [Maksim Kita](https://github.com/kitaisreal)
+* Review Manager: TBD
+* Status: **Awaiting implementation**
+* Implementation: [apple/swift#30242](https://github.com/apple/swift/pull/30242), [apple/swift-evolution-staging#1](https://github.com/apple/swift-evolution-staging/pull/1)
+* Bugs: [SR-8190](https://bugs.swift.org/browse/SR-8190), [SR-6868](https://bugs.swift.org/browse/SR-6868), [SR-11956](https://bugs.swift.org/browse/SR-11956)
+
+## Introduction
+
+Introduce the `CircularBuffer` collection type conforming to the `RandomAccessCollection`, 
+`RangeReplaceableCollection` and `MutableCollection` protocols. With random element 
+access and support for constant back and front element insertion and deletion.
+
+https://en.wikipedia.org/wiki/Circular_buffer
+
+## Motivation
+
+Swift currently does not have collection with both element random access
+and constant O(1) elements back and front insertion and deletion. A good
+usage examples for such collection are queue, deque, fixed length buffers. 
+Such abstractions cannot effectively be build on Array because of O(n) first element deletion.
```

Comment:
Removed `be`, it sounds like an additional word!

```suggestion
Such abstractions cannot effectively build on Array because of O(n) first element deletion.
```